### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by ilyatim**
-To display data, used recyclerView with different viewtypes.
-ViewModel in conjunction with livedata were used to process the activity lifecycle and perform some logic on the data
-Сustom callbacks were used to process clicks on the recycler view elements
-After clicking on the "submit" button, the data is Packed into a json object and displayed as a Toast

### CODE REVIEW

File: app/src/main/java/com/example/testtaskappintheair/ui/SubmitFragment.kt [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        }
+
+        collapsingToolbarLayout?.title = " "
+        view.findViewById<Toolbar>(R.id.submit_fragment_toolbar).setNavigationOnClickListener {
+            activity.finish()
+        }
+
+        mAppBarLayout?.addOnOffsetChangedListener(
+            AppBarLayout.OnOffsetChangedListener { appBarLayout, verticalOffset ->
+                if (appBarLayout.totalScrollRange + verticalOffset < 650) {
```
Comment (rozag) [inline app/src/main/java/com/example/testtaskappintheair/ui/SubmitFragment.kt:155]: This hardcoded value looks scary. I guess things'll break if we launch the app on a device with large screen or just add more items to the list. It's better to handle such cases with some ratio (e.g. `a / b < 0.5`). Let's fix this

File: app/src/main/java/com/example/testtaskappintheair/adapter/RecyclerViewAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            is RecyclerViewCell.CrowdRateDataClass -> ViewType.CROWDRATE
+            is RecyclerViewCell.FeedbackDataClass -> ViewType.FEEDBACK
+            is RecyclerViewCell.ButtonDataClass -> ViewType.BUTTON
+            is RecyclerViewCell.RateWithCheckBoxDataClass -> ViewType.CHECKBOX
+        }
+
+    fun updateAll(newItems: List<RecyclerViewCell>, pos: Int) {
+        items = newItems
+        notifyItemChanged(pos)
+    }
```
Comment (rozag) [inline app/src/main/java/com/example/testtaskappintheair/adapter/RecyclerViewAdapter.kt:35]: This approach works for now, but it's too fragile. Over time we've learnt that any `notifyItem*` methods invocations will lead to crashes, one way or another. We shouldn't perform these things manually. Instead, let's do the following:
- Check out the following sections of our guides: `005-recycler-view.md#diffutil` and `017-kotlin.md#adapter`
- Use [`ListAdapter`](https://developer.android.com/reference/androidx/recyclerview/widget/ListAdapter) instead of our `AbsDiffableListAdapter` here
- `Diffable` interface looks like this and is implemented by our cell classes:
    ```kotlin
    interface Diffable<in T> {
        fun isSame(other: T): Boolean
        fun isContentsSame(other: T): Boolean
    }
    ```
- And we'll be working with `Adapter` in a single place - `Observer` of our `LiveData` with a list of cells - and it'll look like this (`006-view-model-live-data.md#updating-recyclerview-with-livedata` guide):
    ```kotlin
    myViewModel.myLiveData.observe(viewLifecycleOwner) { myWhatever: Type? ->
        // ...
        val adapter = myRecyclerView.adapter
        if (adapter == null) {
            myRecyclerView.adapter = MyAdapter(
                cells,
                // ...
            )
        } else {
            (adapter as MyAdapter).submitList(cells)
        }
    }
    ```

File: app/src/main/java/com/example/testtaskappintheair/ui/SubmitFragment.kt [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        ) {
+            ratingBar.setIsIndicator(checked)
+            viewModel.dataUpdate(
+                pos,
+                rating,
+                checked
+            )
+            if (checked and (rating > 0)) {
+                lifecycleScope.launch(Dispatchers.Main) {
+                    adapter?.updateAll(viewModel.getData(), pos)
```
Comment (rozag) [inline app/src/main/java/com/example/testtaskappintheair/ui/SubmitFragment.kt:71]: `LiveData` is pointless in this implementation. We can replace it with regular fields and it'll work the same. Instead, we should `observe` `LiveData` and let the `ViewModel` decide what should be displayed on the screen. For example, we shouldn't perform `ratingBar.setIsIndicator(checked)` as we do it above. Instead, our cells class should contain this flag and `DiffUtil` will help us to update our `ViewHolder` when data changes.
In other words, we should move towards unidirectional data flow here. 
1. `ViewModel` dispatches some initial data to the `Fragment` via `LiveData`
1. User performs some action, our `ViewModel` is notified about that via some `onSomethingChanged` method
1. `ViewModel` updates it's internal state, rebuilds the list of cells and other view state classes and dispatches them via `LiveData`

File: app/src/main/java/com/example/testtaskappintheair/SubmitViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+    private val resources = getApplication<Application>().applicationContext.resources
+
+    private val headerRate: MutableLiveData<ClassicRateDataClass> = MutableLiveData(
+        createRecyclerCell(
+            resources.getString(R.string.header_first_text_view_experience),
+            0,
+            RateDataType.EXP
+        )
+    )
+    private val items: MutableLiveData<MutableList<RecyclerViewCell>> =
```
Comment (rozag) [inline app/src/main/java/com/example/testtaskappintheair/SubmitViewModel.kt:25]: The content of any `LiveData` should also be immutable. It means that we cannot use `MutableList` here. Instead, we should use simple `List` and our state updates will look like this:
```kotlin
private val _liveData = MutableLiveData<List<...>>()
val liveData: LiveData<List<...>> = _liveData

val oldList = _liveData.value ?: return
val newList = oldList.toMutableList()
// update this mutable newList
_liveData.value = newList
```

File: app/src/main/java/com/example/testtaskappintheair/ui/SubmitFragment.kt [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+    private var recyclerView: RecyclerView? = null
+    private var collapsingToolbarLayout: CollapsingToolbarLayout? = null
+    private var mAppBarLayout: AppBarLayout? = null
+
+    private val onSubmitButtonClickListener = View.OnClickListener {
+        Toast.makeText(
+            context,
+            viewModel.getSubmitData().toJson(),
+            Toast.LENGTH_SHORT
+        ).show()
```
Comment (rozag) [inline app/src/main/java/com/example/testtaskappintheair/ui/SubmitFragment.kt:54]: In real world we'll have to send this data to server, display some progress while we're waiting for a response, and after that display some kind of success message. Let's emulate this behavior:
1. When user taps `Submit` we'll show some `ProgressBar` for a second or two. Note that while this progress is shown, our `RatingBar`s, `EditText`, etc. should be disabled
1. And only after that we will display this `Toast`. Also note, that `Fragment` shouldn't be responsible for this stuff. Our `ViewModel` will perform the network request in real project. So in this case `ViewModel` will decide what to show and when to show it. It'll deliver data to the `Fragment` via `LiveData`, and our `Fragment` will simply render these list or progress states.
1. You can use this `SingleEventLiveData` to deliver this `Toast` text from `ViewModel` to `Fragment`:
```java
/**
 * A lifecycle-aware observable that sends only new updates after subscription, used for events like
 * navigation and Snackbar messages.
 * <p>
 * This avoids a common problem with events: on configuration change (like rotation) an update
 * can be emitted if the observer is active. This LiveData only calls the observable if there's an
 * explicit call to setValue() or call().
 * <p>
 * Note that only one observer is going to be notified of changes.
 */
public class SingleEventLiveData<T> extends MutableLiveData<T> {

    private static final String TAG = "SingleEventLiveData";

    private final AtomicBoolean mPending = new AtomicBoolean(false);

    @MainThread
    @Override
    public void observe(@NonNull final LifecycleOwner owner, @NonNull final Observer<T> observer) {
        if (hasActiveObservers()) {
            MainHelper.log(TAG, "Multiple observers registered but only one will be notified of changes.");
        }

        // Observe the internal MutableLiveData
        super.observe(owner, new Observer<T>() {
            @Override
            public void onChanged(@Nullable final T t) {
                if (mPending.compareAndSet(true, false)) {
                    observer.onChanged(t);
                }
            }
        });
    }

    @MainThread
    @Override
    public void setValue(@Nullable final T t) {
        mPending.set(true);
        super.setValue(t);
    }

    /**
     * Used for cases where T is Void, to make calls cleaner.
     */
    @MainThread
    public void call() {
        setValue(null);
    }

    /**
     * Used for cases where T is Void from bg thread, to make calls cleaner.
     */
    @MainThread
    public void postCall() {
        postValue(null);
    }

}
```

**Review by rozag — CHANGES_REQUESTED** [review]
